### PR TITLE
feat: add element indexing

### PR DESCRIPTION
**Pull request by jamesmporter**

### CODE REVIEW

File: packages/d3fc-webgl/src/program/programBuilder.js [hunk 1]
```diff
@@ -26,7 +27,20 @@ export default () => {
         );
         buffers(context, program, verticesPerElement, count);
 
-        context.drawArrays(mode, 0, count * verticesPerElement);
+        let elementBuffer = null;
+        if (elementIndices != null) {
+            elementBuffer = elementIndices(context, verticesPerElement, count);
+            context.bindBuffer(context.ELEMENT_ARRAY_BUFFER, elementBuffer);
+
+            context.drawElements(
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/program/programBuilder.js]: This doesn't look like the right argument list https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/drawElements

File: packages/d3fc-webgl/src/program/programBuilder.js [hunk 2]
```diff
@@ -26,7 +27,20 @@ export default () => {
         );
         buffers(context, program, verticesPerElement, count);
 
-        context.drawArrays(mode, 0, count * verticesPerElement);
+        let elementBuffer = null;
+        if (elementIndices != null) {
+            elementBuffer = elementIndices(context, verticesPerElement, count);
+            context.bindBuffer(context.ELEMENT_ARRAY_BUFFER, elementBuffer);
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/program/programBuilder.js]: Is this being done twice?

File: packages/d3fc-webgl/src/program/programBuilder.js [hunk 3]
```diff
@@ -10,6 +10,7 @@ export default () => {
     let mode = drawModes.TRIANGLES;
     let buffers = bufferBuilder();
     let verticesPerElement = 1;
+    let elementIndices = null;
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/program/programBuilder.js]: Is prefer all the buffers stayed in buffer builder or moved into here. As keeping them in buffer builder is the smaller change can we go for that for now?

File: packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            return;
+        }
+        if (buffer == null) {
+            buffer = gl.createBuffer();
+        }
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, buffer);
+
+        const bufferArray = [];
+        for (let index = 0; index < count; index++) {
+            bufferArray.push(...data.map(x => x + index * verticesPerElement));
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js]: Am I right in thinking this is only required until we get the ANGLE stuff in?

File: packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    const base = (gl, verticesPerElement, count) => {
+        if (!changed) {
+            return;
+        }
+        if (buffer == null) {
+            buffer = gl.createBuffer();
+        }
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, buffer);
+
+        const bufferArray = [];
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js]: Can we assign directly to a Uint16Array to save the latter copy operation?

**Review by chrisprice — CHANGES_REQUESTED** [review]
Nice one on getting to the bottom of why this wasn't working. Can you roll it out to the other sites types as well?

File: packages/d3fc-webgl/src/program/programBuilder.js (hunk 1 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/program/programBuilder.js]: Good point. I'll change the offset to `0`.

File: packages/d3fc-webgl/src/program/programBuilder.js (hunk 2 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/program/programBuilder.js]: Yes, it doesn't need to be. I've removed this occurrence.

File: packages/d3fc-webgl/src/program/programBuilder.js (hunk 3 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/program/programBuilder.js]: Agreed - I'll change that over.

File: packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js (hunk 4 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js]: Yes

File: packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js (hunk 5 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js]: Done.

**Comment by jamesmporter** [comment]
@chrisprice I've addressed all of your comments.

File: packages/d3fc-webgl/src/program/programBuilder.js [hunk 6]
```diff
@@ -26,7 +26,16 @@ export default () => {
         );
         buffers(context, program, verticesPerElement, count);
 
-        context.drawArrays(mode, 0, count * verticesPerElement);
+        if (buffers.elementIndices() != null) {
+            context.drawElements(
+                mode,
+                buffers.elementIndices().data().length * count,
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/program/programBuilder.js]: `buffers.elementIndices().data().length` === `verticesPerElement`?

File: packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    const base = (gl, cornerCount, elementCount) => {
+        if (!changed) {
+            return;
+        }
+        if (buffer == null) {
+            buffer = gl.createBuffer();
+        }
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, buffer);
+
+        const vertexCount = data.length;
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js]: Is this what we'd normally call this `verticesPerElement`? Should we pass it in here and assert that data is the right length?

**Review by chrisprice — CHANGES_REQUESTED** [review]
A few minor comments on the fallout from the changes. I'm keen to get #1407 in first if we can so that we can test the interaction I mentioned in a comment on the PR.

File: packages/d3fc-webgl/src/program/programBuilder.js (hunk 6 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/program/programBuilder.js]: No.

`verticesPerElement` = number of corners defined.
`elementIndices length` = number of vertices defined (via indices).

I found this quite confusing when writing the code so perhaps we should change one of these names to clarify what is going on.

File: packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js (hunk 7 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/elementIndicesBuilder.js]: Unfortunately no, as per my comment above.

Also, due to the *ANGLE work, this line has been removed.

File: packages/d3fc-webgl/src/program/programBuilder.js (hunk 6 above)
Comment (chrisprice) [reply packages/d3fc-webgl/src/program/programBuilder.js]: Of course. I think we should now be able to remove `verticesPerElement`, although let's do that in another PR.

**Review by chrisprice — APPROVED** [review]

**Pull request merged**